Add basic mychevy component

sdague opened the pull request:
This implements a component using the mychevy library (which utilizes
selenium to webscrape the mychevy website for onstar for your
car). The architecture works by having a background thread doing
periodic polling of the website, and updating the sensors when new
data is found.

This requires rather more setup than most platforms, as you need
working selenium. Instructions will be provided on the component
list. All the sensors are spawned and coordinated from a single "hub"
as they are really just attributes of the same web scraping session.

## Description:


**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4410

## Example entry for `configuration.yaml` (if applicable):
```yaml
mychevy:
  username: email
  password: password
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

sdague committed:
mychevy: added basic mychevy component

This implements a component using the mychevy library (which utilizes
selenium to webscrape the mychevy website for onstar for your
car). The architecture works by having a background thread doing
periodic polling of the website, and updating the sensors when new
data is found.

This requires rather more setup than most platforms, as you need
working selenium. Instructions will be provided on the component
list. All the sensors are spawned and coordinated from a single "hub"
as they are really just attributes of the same web scraping session.

balloob commented:
Do you think that this will remain a sensor only platform or do you think that it will move in the future to a component that handles auth and a sensor, binary sensor and switch platforms? If the latter, let's move the auth to a `mychevy` component.

sdague commented:
I probably should actually represent at least one of these as a binary sensor. I'm happy to move it to a topline component to be consistent there. I can do that tomorrow.

tinloaf commented:
Regarding the fact that you spawn your own process: If you implemented all the scraping (i.e., spawning selenium, authenticating, …) inside your `update` method, it should be executed inside a thread executor, i.e., not block the rest of HA. I'm not saying that this is strictly the better solution, just want to point out that this is a possibility.

I guess this is mainly weighting the advantages of having a persistent selenium process (speed up b/c of not having to reauthenticate etc) against the disadvantages of having one more thread lying around (more RAM usage, debugging becomes harder, …)

sdague commented:
@tinloaf in my local testing this approach did not block any of the rest of HA. If you've got an example in the home assistant code of the other approach, that would be great, and I can look into it. I had some early attempts to make this asyncio directly, but selenium just doesn't play well with that.

The overhead of the selenium start and login is trivial (2 seconds) compared to the 60+ seconds of refresh time it takes to get the actual data over the onstar link. So doing the login repeatably instead of having to catch the cases of credential timeout seemed like the best bet.

sdague commented:
This is now implemented as a component, though the logic of who owns the sensors still pretty much works the same way as before. I was trying to avoid popping up a bunch of sensors with no values to the user, and instead am using a status sensor to figure out whether or not the platform is behaving. The level of outage that the mychevy website has is really high (like 20% high, it's not been working all morning), so users are going to need an indicator of whether anything is even expected to be working.

This is a little tricky with the sensor discovery, which wants to add sensors when it gets there. I instead just have it pass back the add_devices method so once there is some real data the sensors will pop up. It's a little odd, but I wasn't sure what better options to run with.

MartinHjelmare reviewed (commented):
I think you should set up the sensors normally and use a typical data class in place of the hub class. There are many examples of that. Ie let the sensors drive the update of the client api data class, which has a throttle. You can have the sensors be unavailable when there are connection problems.

Threading interaction with asyncio is not a problem as long as you use the home assistant sync api. Ie `setup` for the component and `setup_platform` + `update` for the platform.

sdague commented:
@MartinHjelmare without the background thread driving updates, how do you handle sensor initiated requests for update to something which is an up to 3 minute blocking call? I guess that's the part I wasn't seeing good examples of in the codebase, and I threw out a few other approaches because it was blocking the main thread.

Specific pointers in the existing code base would be appreciated, both on that issue, and one what's considered best practice on the data class pattern.

MartinHjelmare commented:
Ah ok I didn't understand it was blocking for that amount of time. Yeah then it's best to have a persistent separate thread for the update if asyncio is not possible.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,343 @@
+    def __init__(self, connection):
+        """Initialize sensor with car connection."""
+        self._conn = connection
+        self.hass = connection.hass
```
Home assistant will set `hass` on the entity when it has been added. So remove this.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,343 @@
+        self._car = data
+        for sensor in self.sensors:
+            sensor.car = data
+            sensor.schedule_update_ha_state()
```
Instead of calling this directly on each sensor, send a signal to each sensor via our dispatch helper. The target for the signal is the method on the entity that schedules the update.

To avoid a race where the update is called before the entity has been added to home assistant, connect the dispatcher in `async_added_to_hass` coroutine method on the entity.

sdague replied:
Do you have an example of that? Using this schedule call seems really straight forward, and I'm not sure I've got a clear picture of how the code would look to basically build this myself. Any examples would be appreciated.

MartinHjelmare replied:
It's precarious to call `schedule_update_ha_state` based on an outside home assistant event, in this case when the car is updated, since you don't know if the entity has been added to home assistant yet. That's why it's good to use the dispatcher. The dispatcher allows you to decide when to connect the signal with the target. That should happen after the entity has been added to home assistant, ie in the `async_added_to_hass` method.

See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/alarm_control_panel/envisalink.py#L97-L109

Note that you can implement the async version of the state update scheduling, since the dispatcher will schedule a job already on the event loop.

Since you always want to update all sensors you can have the same signal for all sensors.

Then you send the signal with `dispatcher_send`:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/dispatcher.py#L56

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,343 @@
+
+    def __init__(self, connection):
+        """Initialize sensor with car connection."""
+        self._state = MYCHEVY_UNKNOWN
```
Init state as `None`. The entity base class will handle unknown state.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,343 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {"last_update": self._last_update}
```
Time stamp is not allowed in the state attributes. Only a time change should not create a state update.

sdague replied:
ack

MartinHjelmare reviewed (commented):
What if a user has more than one car? Depending on how Chevy handles that on their site, it could be better to separate the sensors from the car and just let the car hold the state. The sensors would then instead fetch their state on the car when the car signals the need.

See other comments below.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,15 @@
+    hub = hass.data[mychevy.DOMAIN]
+    hub.register_add_devices(add_devices)
+
+    return True
```
Remove this. This function shouldn't return anything.

sdague replied:
ack

sdague committed:
mychevy: only poll every 30 minutes

sdague committed:
mychevy: update sensors

sdague committed:
mychevy: better error handling

sdague committed:
mychevy: tweaking for refactor

sdague commented:
The user has more than one car issue is unsolved right now. This "library" was only buildable with a lot of time in chrome inspector to understand what the website looks like. If someone shows up with that issue I'll try to get some web page dumps from them to get them working.

sdague committed:
mychevy: bump version to handle odometer > 1000

sdague committed:
mychevy: great sensor reorg

sdague committed:
mychevy: add binary sensors

sdague committed:
mychevy: bump mychevy requirement

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+    @property
+    def icon(self):
+        """Return the icon."""
+        if self._icon_func is not None:
```
Just check if this sensor is a battery sensor, eg:
```py
if self._attr == 'percent':
    return icon_for_battery_level(self._state)
return self._icon
```

sdague replied:
Nice! I didn't realize that was a thing.

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+from logging import getLogger
+
+from homeassistant.components.mychevy import (
+    EVBinarySensorConfig, DOMAIN,
```
This sensor belong to the binary sensor domain. So don't overwrite the domain by importing the mychevy domain with the same name. Do:
```py
from homeassistant.components.mychevy import (
    DOMAIN as MYCHEVY_DATA)
```

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+"""Support for MyChevy sensors."""
```
Please update the module docstring to follow the regular format with a link to the docs:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/luftdaten.py#L1-L6

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+_LOGGER = getLogger(__name__)
+
+SENSORS = [
+    EVBinarySensorConfig("Plugged In", "plugged_in", "plug")
```
The ususal way to do this in home assistant is to use a dict holding a list of name, unit, icon etc. See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/sensor/luftdaten.py#L38-L44

I don't think you need to create a class for the config. I like the verbosity of attribute access that a class gives but I think we should try to stick within the mold as long as the mold will fit. If you want to keep the verbose attribute access you can create a nested list of dicts, instead of a nested list of lists.

```py
SENSORS = [
    {'name': 'EVCharge', 'attr': 'percent', 'unit_of_measurement': '%', 'icon': 'mdi:battery'}
   ...
]
```

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+
+    email = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    if hass.data.get(DOMAIN) is None:
```
Component setup will only ever happen once, so you can remove this check.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+    async_add_job.
+    """
+
+    def __init__(self, client, hass, **kwags):
```
`**kwargs`

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+
+    def __init__(self, client, hass, **kwags):
+        """Initialize MyChevy Hub."""
+        super(MyChevyHub, self).__init__()
```
This class doesn't seem to inherit from anything but `object`.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+            self.car = self._client.data()
+            self.status.success()
+        except Exception:
+            _LOGGER.exception(
```
We usually don't want to log with exception level in home assistant, error level is preferred.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+            self.update()
+            self.car = self._client.data()
+            self.status.success()
+        except Exception:
```
Is it impossible to narrow this down? If not disable the pylint warning for broad except.

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+    ENTITY_ID_FORMAT, BinarySensorDevice)
+from homeassistant.util import (slugify)
+
+_LOGGER = getLogger(__name__)
```
We usually write:
```
_LOGGER = logging.getLogger(__name__)
```

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+                "This probably means the OnStar link is down again")
+            self.status.error()
+
+    def get(self, attr, default=None):
```
`default` is never changed when this method is called. Is it needed?

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+    def update(self):
+        """Update call that is safe to call from async_add_job."""
+        try:
+            self.update()
```
Why call itself here?

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+
+
+SENSORS = [
+    EVSensorConfig("Mileage", "mileage", "miles", "mdi:speedometer"),
```
See comment for binary sensor.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+    def __init__(self, connection):
+        """Initialize sensor with car connection."""
+        self._state = None
+        self._last_update = dt.now()
```
This isn't used.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+        self.hass = hass
+        self.car = None
+        self.status = None
+        self.sensors = []
```
This isn't used.

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+        """Initialize MyChevy Hub."""
+        super(MyChevyHub, self).__init__()
+        self._client = client
+        self.hass = hass
```
This isn't used.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+    def __init__(self, connection, config):
+        """Initialize sensor with car connection."""
+        self._conn = connection
+        connection.sensors.append(self)
```
This can be removed.

sdague replied:
A yep

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+        """Initialize sensor with car connection."""
+        self._state = None
+        self._last_update = dt.now()
+        self._conn = connection
```
This isn't used. You could refactor this entity to fetch the connection status from `self._conn` instead of the connection instance calling either of the methods `success` or `error` on this entity. The former might be more natural and in line with the rest of the sensors here and also the rest of the sensors in home assistant.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+        """Update chevy data."""
+        # safe_update takes 2 - 3 minutes to run, it's fine to just
+        # get it started, then update with whatever we find.
+        self.hass.async_add_job(self._conn.update)
```
If this takes 2-3 min to complete, it will occupy a worker pool thread for that time. I'm not sure if that is ok. I think we need a comment from core on this.

sdague replied:
Yeh, I'm not either. It is easy enough to build my own dedicated thread. I'm happy to go whatever way people would rather here.

balloob replied:
I am also in favor of a dedicated thread 👍 

Since the update is so irregular, it would probably be more efficient to tell the entities when an update is available instead of having Home Assistant ask the entities every 15 seconds to check with the hub if an update is available.

To do that:

 - All entities disable polling (return `False` from `should_poll` property on the entity)
 - All entities call `hass.helpers.dispatcher.dispatcher_connect('mychevy', self.async_update_ha_state)` to listen for updates. (or `async_dispatcher_connect` if in async context
 - Thread will schedule itself. When an update succeeds, call `hass.helpers.dispatcher.dispatcher_send('mychevy')`

balloob replied:
(the [dispatcher](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/dispatcher.py) is a mini event bus that is for intra-integration communication)

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+]
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Since no I/O is performed in this function you can switch to `async_setup_platform`.

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,68 @@
+    def __init__(self, connection, config):
+        """Initialize sensor with car connection."""
+        self._conn = connection
+        connection.sensors.append(self)
```
This can be removed, since `connection.sensors` is not used in the `connection` instance.

MartinHjelmare reviewed (dismissed):
I like the refactor, but I'm concerned about occupying a worker thread for 2-3 min. I think someone in core needs to comment on that.

sdague commented:
@MartinHjelmare agreed on the 2 - 3 minutes thing. It's easy enough to bring back in the dedicated worker thread. I'm fine either way. Mostly this became educational in how one would do an asyncio version of it.

The good news, I've got both the old and new versions running in the house, and they seem to work well (clarifying "well" to include the fact that it's a glorious day when the chevy website stays up for more than 6 hours in a single day... )

sdague commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,155 @@
+from datetime import datetime as dt
+
+from homeassistant.components.mychevy import (
+    EVSensorConfig, DOMAIN, MYCHEVY_ERROR, MYCHEVY_SUCCESS,
```
Ack. Yep that slipped through in the refactor.

balloob commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,121 @@
+            _LOGGER.exception(
+                "Error updating mychevy data. "
+                "This probably means the OnStar link is down again")
+            self.status.error()
```
Would you expect to try sooner when there is an error? (not possible with `Throttle`)

sdague replied:
Actually not. Their site doesn't recover from fail states very quickly, and I'm a little concerned that faster polling during errors might make things worse.

sdague commented:
@balloob thank you for the look and the feedback! I'll do a respin hopefully this weekend.

sdague committed:
mychevy: use dispatcher

Instead of directly modifying the sensors, this lets us use a
dispatcher to have sensors pull information from the car object when
there is a relevant update for them.

sdague commented:
Ok, new stack is pushed. This is the more raw set of commits to make it easier for me to change things for any needed set of feedback. It uses the thread and dispatcher.

The dispatcher confused me a little, because I had assumed that calling async_update_ha_state would call my update method, but that didn't seem to be the case. So I'm routing through an update_callback instead. If I'm doing that wrong let me know.

sdague committed:
mychevy: remove from coverage

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,163 @@
+        self._icon = config.icon
+        self._state = None
+
+        self.entity_id = ENTITY_ID_FORMAT.format(
```
You can remove this. It will be done by the core.

sdague replied:
core will automatically name sensors by platform? I didn't think I saw that behavior when I tried.

MartinHjelmare replied:
Yeah, binary sensors will have entity_id `binary_sensor.name` and sensors will have entity_id `sensor.name` eg.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,163 @@
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        self.hass.helpers.dispatcher.async_dispatcher_connect(
+            DOMAIN, self.async_update_ha_state)
```
I think you want to connect to `async_update_callback`.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,163 @@
+        """Return the name."""
+        return self._name
+
+    @asyncio.coroutine
```
This should be a callback.

sdague replied:
If it calls async_schedule_update_ha_state, doesn't it need to be a coroutine?

MartinHjelmare replied:
No, that method is a callback that will schedule the update on the event loop.

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,79 @@
+        self.hass.helpers.dispatcher.async_dispatcher_connect(
+            DOMAIN, self.async_update_callback)
+
+    @asyncio.coroutine
```
This should be a callback.

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,79 @@
+        """Update state."""
+        if self._conn.car is not None:
+            self._is_on = getattr(self._conn.car, self._attr, None)
+            yield from self.async_update_ha_state()
```
Just call `async_schedule_update_ha_state`. That will schedule the state update.

sdague replied:
ack

sdague committed:
mychevy: dedicated constants for dispatch signals

This makes the dispatch signals dedicated topics, and fixes updating
the state on the sensors so that they are correctly updated.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,163 @@
+        """Update state."""
+        if self._conn.car is not None:
+            self._state = getattr(self._conn.car, self._attr, None)
+            yield from self.async_update_ha_state()
```
Just call `async_schedule_update_ha_state`.

sdague replied:
ack

MartinHjelmare reviewed (changes requested):
Ok. The structure is ok now. There are still remaining comments from my previous review. Ive removed the comments that aren't relevant anymore. I've added more comments below.

sdague commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,132 @@
+    starts.
+    """
+
+    def __init__(self, client, hass, **kwags):
```
ack

sdague commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,132 @@
+
+    def __init__(self, client, hass, **kwags):
+        """Initialize MyChevy Hub."""
+        super(MyChevyHub, self).__init__()
```
ack

sdague commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,163 @@
+            self._state = MYCHEVY_SUCCESS
+        self.schedule_update_ha_state()
+
+    def error(self):
```
ack

sdague committed:
mychevy: updated with comments from martinhjelmare

sdague committed:
mychevy: set battery icon based with helper function

sdague committed:
Address additional review feedback

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,85 @@
+from homeassistant.core import callback
+from homeassistant.util import (slugify)
+
+_LOGGER = getLogger(__name__)
```
Please follow the same pattern as other modules.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,85 @@
+    @callback
+    def async_update_callback(self):
+        """Update state."""
+        if self._conn.car is not None:
```
When can this happen?

sdague replied:
If the initial update is a failure, there will be no data object there

MartinHjelmare replied:
:+1:

MartinHjelmare commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,85 @@
+"""Support for MyChevy sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/mychevy/
```
Wrong url.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/mychevy.py:
```diff
@@ -0,0 +1,133 @@
+    import mychevy.mychevy as mc
+
+    config = base_config.get(DOMAIN)
+    _LOGGER.debug('Received configuration: %s', config)
```
I don't think you should log the credentials.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,164 @@
+    EVSensorConfig("EVCharge", "percent", "%", "mdi:battery")
+]
+
+_LOGGER = getLogger(__name__)
```
Same as for binary sensor.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,164 @@
+from logging import getLogger
+
+from homeassistant.components.mychevy import (
+    EVSensorConfig, DOMAIN, MYCHEVY_ERROR, MYCHEVY_SUCCESS,
```
Don't overwrite the sensor domain.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,164 @@
+    def success(self):
+        """Update state, trigger updates."""
+        if self._state != MYCHEVY_SUCCESS:
+            _LOGGER.info("Successfully connected to mychevy website")
```
Debug.

sdague replied:
ack

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,164 @@
+    @property
+    def icon(self):
+        """Return the icon."""
+        if self._icon == "mdi:battery":
```
It's a bit weird that you define a battery icon in the config and then overwrite that. I'd use the `_attr` as I'd to check icon for battery.

MartinHjelmare commented on homeassistant/components/sensor/mychevy.py:
```diff
@@ -0,0 +1,164 @@
+"""Support for MyChevy sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/mychevy/
```
Wrong url.

sdague replied:
ack

sdague commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,85 @@
+from homeassistant.components.binary_sensor import (
+    ENTITY_ID_FORMAT, BinarySensorDevice)
+from homeassistant.core import callback
+from homeassistant.util import (slugify)
```
ack

sdague commented on homeassistant/components/binary_sensor/mychevy.py:
```diff
@@ -0,0 +1,85 @@
+from logging import getLogger
+
+from homeassistant.components.mychevy import (
+    EVBinarySensorConfig, DOMAIN, UPDATE_TOPIC
```
ack

sdague committed:
Address additional review comments

MartinHjelmare reviewed (approved):
Looks good! I'm not going to fight about the sensor attribute config thing. But I hope you see my point about reusing patterns makes it easier to maintain the whole code.

sdague commented:
@MartinHjelmare sure, I didn't invent the pattern from whole cloth though, WUnderground module uses it as well - https://github.com/home-assistant/home-assistant/blob/6ef02e8ede622076729fae83bc4e111450d5adee/homeassistant/components/sensor/wunderground.py#L39 which is where I tripped over it in the code.

MartinHjelmare commented:
👌 yeah I shouldn't blame you for tripping over gold in the woods. :smile:

Good work! :tada:
